query.cc

```diff
@@ -412,9 +412,14 @@ static bytes_opt merge_singular_results(bytes_opt r1, bytes_opt r2, forward_requ
     throw std::runtime_error("unknown reduction type");
 }
 
-void forward_result::merge(const forward_result& other, const std::vector<forward_request::reduction_type>& reduction_types) {
+forward_result& forward_result::merge(const forward_result& other, const std::vector<forward_request::reduction_type>& reduction_types) {
     if (query_results.empty()) {
-        query_results.resize(other.query_results.size());
+        // Merging empty result (this) with a non empty one (other) requires copying the other
+        query_results = other.query_results;
+        return *this;
+    } else if (other.query_results.empty()) {
+        // Merging non empty result (this) with an empty one (other) is a noop
+        return *this;
```

Comment:
Because the fix is intermingled with the simplification, it's harder to understand either.